### PR TITLE
[Drivers] Support InfiniBand drivers installation

### PR DESCRIPTION
**Pull request by ydye**
Add MLNX OFED 4.2-1.2.0.0 and nvidia-peer-memory_1.0.5 into each drivers image.

### CODE REVIEW

File: src/drivers/build/drivers-384.111.dockerfile [hunk 1]
```diff
@@ -48,7 +48,36 @@ RUN apt-get -y update && \
         python-pip \
         realpath \
         gawk \
-        module-init-tools && \
+        module-init-tools \
```
Comment (xudifsd) [inline src/drivers/build/drivers-384.111.dockerfile]: I really think we should have only one template file, and instantiated version specific file in build-pre.sh

File: src/drivers/build/drivers-384.111.dockerfile [hunk 2]
```diff
@@ -57,13 +86,39 @@ RUN apt-get -y update && \
...
 RUN wget --no-verbose http://us.download.nvidia.com/XFree86/Linux-x86_64/$NVIDIA_VERSION/NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run && \
     chmod 750 ./NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run && \
     ./NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run --extract-only && \
     rm ./NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run
 
+RUN echo "wget -q -O - http://www.mellanox.com/downloads/ofed/MLNX_OFED-$OFED_VERSION/$MLNX_OFED_STRING.tgz | tar xzf -" && \
+    wget -q -O - http://www.mellanox.com/downloads/ofed/MLNX_OFED-$OFED_VERSION/$MLNX_OFED_STRING.tgz | tar xzf - && \
+    echo "wget -q -O - http://www.mellanox.com/downloads/ofed/nvidia-peer-memory_1.0.5.tar.gz | tar xzf -" && \
+    wget -q -O - http://www.mellanox.com/downloads/ofed/nvidia-peer-memory_1.0.5.tar.gz | tar xzf - && \
+    git clone https://github.com/NVIDIA/gdrcopy.git
```
Comment (xudifsd) [inline src/drivers/build/drivers-384.111.dockerfile]: add `--depth=1` if we only need the master branch. Or we can download tar from github directly if we do not need version control functionality.

File: src/drivers/build/drivers-384.111.dockerfile [hunk 3]
```diff
@@ -57,13 +86,39 @@ RUN apt-get -y update && \
...
+
+ENV MLNX_OFED_STRING=MLNX_OFED_LINUX-${OFED_VERSION}-${OS_VERSION}-${ARCHITECTURE}
 
 RUN wget --no-verbose http://us.download.nvidia.com/XFree86/Linux-x86_64/$NVIDIA_VERSION/NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run && \
     chmod 750 ./NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run && \
     ./NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run --extract-only && \
     rm ./NVIDIA-Linux-x86_64-$NVIDIA_VERSION.run
 
+RUN echo "wget -q -O - http://www.mellanox.com/downloads/ofed/MLNX_OFED-$OFED_VERSION/$MLNX_OFED_STRING.tgz | tar xzf -" && \
+    wget -q -O - http://www.mellanox.com/downloads/ofed/MLNX_OFED-$OFED_VERSION/$MLNX_OFED_STRING.tgz | tar xzf - && \
```
Comment (xudifsd) [inline src/drivers/build/drivers-384.111.dockerfile]: maybe untar the file in installation script? the drivers image is already 3G large.

Comment (ydye) [reply src/drivers/build/drivers-384.111.dockerfile]: @xudifsd  We should consider some inter-networking environment may can't access to outer-networking. So we have to prepare all downloading and installation dependency in the docker image.

**Comment by ydye** [comment]
@xudifsd  Maybe we could add a new base image of drivers to gather the common part of each image, and then build different version's driver from the base image.

File: src/drivers/build/drivers-384.111.dockerfile (hunk 3 above)
Comment (xudifsd) [reply src/drivers/build/drivers-384.111.dockerfile]: yes, I know, so I only ask not to untar.

**Comment by xudifsd** [comment]
This remains me if we should put ib driver and nvidia driver in same service. If user do not need ib drivers, he should not download ib part.

**Comment by ydye** [comment]
Ib part's installation will have a dependency on Nvidia drviers. Because GDR need GPU drivers to be installed firstly.

File: src/drivers/build/drivers-384.111.dockerfile (hunk 3 above)
Comment (ydye) [reply src/drivers/build/drivers-384.111.dockerfile]: Got it, I will try to optimize it.

**Comment by fanyangCS** [comment]
https://github.com/Microsoft/pai/issues/1732

**Comment by ydye** [comment]
@xudifsd after investigating, I found the saved space is very little. And I think it's not worth to refactor the dockerfile with such little gains.

**Comment by ydye** [comment]
```
112K    ./nvidia-peer-memory-1.0
440K    ./gdrcopy
149M    ./MLNX_OFED_LINUX-4.2-1.2.0.0-ubuntu16.04-x86_64
149M    .
```

**Comment by xudifsd** [comment]
you mean using base image?

**Comment by ydye** [comment]
> you mean using base image?

I think the base image is a concept for human. Actually for drivers-390.25 and drivers-410.73, they have a lot of same parts, so they share the same layer of image in system. 
I think the benefits of writing a based image is comfortable for developer to change the common part code rather then saving space.

**Comment by xudifsd** [comment]
Oh, you misunderstood, I was asking what you have investigated? Using base image or not using untar in build.

**Comment by ydye** [comment]
I think untar's gain is little.

**Comment by xudifsd** [comment]
Ok. Then will you use base image or we merge this?

**Comment by ydye** [comment]
I still need test the installation performance, later merge it. And the base image design can be done in another PR.

**Comment by xudifsd** [comment]
ok, ping me if you need review.

**Pull request merged**